### ISSUE TITLE
To what degree should xmldom support DTD processing support for Entity/ EntityReference?

### ISSUE THREAD

**Issue by pmahend1** [xmldom/xmldom#117]
I dont know if the parser supports extracting `Entity` and `EntityReference` but they are not in my case.

Below is the XML (reference https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlreader.read?view=netcore-3.1#examples)

```xml
<?xml version="1.0"?>
<!-- This is a sample XML document -->
<!DOCTYPE Items [
  <!ENTITY num "123">
]>
<Items ok="yes" okay="okay">
               <Item>Test with an entity: &num;</Item>
  <Item>Test with a child element <more /> stuff</Item>
  <Item>Test with a CDATA section <![CDATA[<456>]]> def</Item>
  <Item>Test with a char entity: &#65;</Item>
  <!-- Fourteen chars in this element.-->
  <Item>1234567890ABCD</Item>
</Items>
```

I dont get Entity node from the parser.
If you see the example code there for `DocumentType`  it parses Entity in the above example as is , but `DomParser` isnt parsing it.

```C#
case XmlNodeType.DocumentType:
   Console.Write("<!DOCTYPE {0} [{1}]", reader.Name, reader.Value); //reader.value is not fetched by DomParser
   break;
```

Also why significant whitespace with next line are parsed as Text codes? They probably dont qualify to be termed as Text even though its similar but, I think they may have to be emitted as WhiteSpace or Significant whitespace Nodes.

See the C# NodeType enum which does include `SignficantWhiteSpace` here https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlnodetype?view=netcore-3.1

Other Example of DTD : https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlreadersettings.dtdprocessing?view=netcore-3.1#input

**Comment by karfau** [comment]
Thx for raising this issue.
I'm aware that there is at least one issue with processing `DOCTYPE`s and wanted to create a ticket for that. But I first wanted to have a better understanding "what is going on"/"how bad it is".

As soon as I'm done with my current issue, I'll dig into that topic and hopefully be able to reply to your questions.

The only thing that confuses me a bit in your post is the .NET syntax.
Can you maybe elaborate on your setup, so we can better understand how this JS library has an impact on .NET code? Thx

**Comment by pmahend1** [comment]
@karfau C# code is irrelevant for the working of DomParser.  The issue is, I am expecting `<Entity num "123" />` to be parsed but it's not. I think I am expecting the DomParser to work just like C# XMLReader.  So C# is for reference.

**Comment by karfau** [comment]
OK, thank you for the clarification.

I must say: Although this library has been around for a while and is used by many it also has been abandoned for years and is still work in progress (hence the version starts with `0.`). So expecting the same as from some other mature lib or part of a framework, might be a bit to much.

That being said, I'm looking forward to start looking into it soon.

**Comment by pmahend1** [comment]
@karfau Yes , I understand may be I am expecting too much. I was in search for a package does similar parsing as C# `XmlReader`, this package is closest one to my search. I looked at the _lib/dom-parser_ code, looks like its implemented by `SaxParser`. Also I noticed we can pass `mimetype` as `html`, I will delve too and see if i can get it working or if Sax parser can be improved.

**Comment by karfau** [comment]
I started looking into it and there is more research to do, but I currently believe that the support for `!doctype` is very limited, it might even just support single line ones...

As soon as #112 landed I'll create a branch and start adding some tests, that shows how far the support goes.

**Comment by pmahend1** [comment]
@karfau I also did some research on this and tried various parsers available in NPM repository. Yes the support for DTD is very limited to my surprise. We can get the missing text here through a SaxParser (or SaxEs parser library) but it wont be a DOM and we cant get children of a node . On the other hand DOM parser available in JS doesnt seem to parse this DTD , so far I have seen they dont get `Entity` reference text within `DocType` which is a bummer.

**Comment by karfau** [comment]
Thx for the research.
I think there is lots of stuff to implement here, and since we are not sure if we want to keep this own sax parser implementation in xmldom or switch to something else (see #55) I'm not sure it's the first thing we should work on.
But I think it's totally worth to correctly parse multi line doctypes, even if we just pass the whole content to the DOMHandler as a string in a first step. So that the resulting DOM makes a bit more sense.

@brodybits what do you think about this?

**Comment by pmahend1** [comment]
@karfau yes, its lot of work considering there doesn't seem to be a complete implementation of DocType/DTD in opensource npm.

I think [saacs/sax-js](https://github.com/isaacs/sax-js) is a good library if we want to mix Sax just to get DTD/Doctype outer XML.

A possible solution I can think of is to use DocParser as is and use sax-js to get DocType although I am not sure if maintaining hierarchy is still possible with that approach.

The reason I started to dig into all these parsers is that I have built custom parser on .Net which I am using in VSCode extension through `electron-edge-js`  and there was problem with Mac node binaries not being compiled or found by default. I was able to solve that issue yesterday.  So solution through xmldom is not critical for me now.  I will try to get more information on other available parsers though.

**Comment by karfau** [comment]
Thx for the update.

**Comment by karfau** [comment]
I did some research to better understand the scope of DTD, and came to the following conclusions:
• There is a lot of stuff in DTD syntax that is about the question whether the XML is valid. I believe the "XML validity" is outside of the scope of this library (at least for a v1.0) and since there are quite some limitations in DTD I believe more people are using XSD nowadays.
• The part that might or maybe even should be relevant for xmldom are entities defined in DTD. Since they have an impact on parsing the document.
• default values for attributes could also be considered relevant, but I believe they should be opt-in anyways.
• the most tricky part are "external" definitions that can either be provided "locally" or "publicly". I believe supporting that would best be possible by providing a new option that is responsible for resolving those. This way its easy for people to tweak it to their specific needs.

It could make sense to create a milestone or label for this topic to organize related issues. As soon as it exists I could create some issues that describe a path towards some features.

**Comment by brodybits** [comment]
Thanks for all of the info and research. I have added the labels and added this issue to 0.5.0 milestone for now (no promise).

My one major doubt is that this seems like a possible divergence away from the web APIs that xmldom was originally designed to provide. It seems to me like xmldom is acting like a "ponyfill", as I said in #119, and I think this should be documented.

I wonder if we should consider some improved API layering, as discussed in #54 for example.

**Comment by karfau** [comment]
Yes, we should check what the browser API is and how it treats DTD and only extend support in that way. (I wonder to what extend jsdom is offering this API...)

**Comment by karfau** [comment]
At the current point in time the whole DTD topic is not so much of a priority so I would rather consider it for something that needs to be finally decided when we target 1.0, so I changed the milestone accordingly.

(The research on how far browsers and jsdom support this stuff has not been done and would still be helpful.)

**Comment by niklasl** [comment]
I did a simple patch for basic cases like this (see the draft PR https://github.com/xmldom/xmldom/pull/367).

Both Firefox and Chromium-based web browsers handle these examples using their respective [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) implementations.

(I haven't tried jsdom here. It doesn't handle the related case of handling entities in XHTML though (see https://github.com/jsdom/jsdom/issues/3266).)

**Comment by shunkica** [comment]
Is there any way to get the internal subset part of a  DocumentType node?
I am trying to create a canonicalization function, but xmldom swallows up the internal subset with no apparent way to retrieve it.

For example:
```
<!DOCTYPE doc [<!ATTLIST dummy attr CDATA "default">]>
```
turns into
```
<!DOCTYPE doc>
```

**Comment by karfau** [comment]
@shunkica you mean that there is no way of receiving the raw data from the DOM API?
I'm not aware there is anything in the specs regarding that, but we should of course be able to come up with a non specified way of passing and potentially also preserving the raw data when serializing the doctype.

**Comment by karfau** [comment]
> It would be even better if xmldom parsed it for me :)
> DTD is important for proper c14n

Implementing any parsing/handling for that is totally outside of the scope of what I'm capabale of on multiple levels (knowledge, resources, willingness). But if you want to join as a maintainer (and ideally also bring in more people) to foccus on that part (including test coverage) I would be very happy about it.

But keep in mind that there is the topic of replacing the current sax parser (see #55), and all other XML parsers I have seen do not handle that piece, so even if you manage to pull this off, it might need to be ported to whatever we will replace the sax parser with.
(In which case it might become a dependency which could also be a `@xmldom/???` fork of something existing...)

**Comment by shunkica** [comment]
I would like to help but I do not feel familiar enough with this subject matter to assume such a role.
I can share my current method for parsing the DOCTYPE content into  `publicId`, `systemId `and `internalSubset`, but it is currently a very naive approach.

**Comment by shunkica** [comment]
I have looked into the code responsible, and it looks like you are already setting the `publicId` and the `systemId`.
So the only thing missing is the `internalSubset`. It was added to the DocumentType node in  [XML DOM L2 Core](https://www.w3.org/TR/DOM-Level-2-Core/#core-ID-412266927).

**Comment by karfau** [comment]
With #498 on the finish line xmldom provides a basis for DTD parsing by preserving the DOCTYPE internalSubSet and tons of "grammar" regular expressions to parse them.

It solves tons of issues around parsing/skipping over the doctype, so it's a solid foundation for the future, and will be released as art of 0.9.0 (and related pre-releases).

So I decided to mark this issue as "done".

Feel free to contribute to extend this functionality or open a more specific issue/PR for what exactly you want to request/contribute to extend it.
